qiskit/passmanager/propertyset.py

```diff
@@ -10,15 +10,54 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-"""A property set is maintained by the pass runner.
+"""A property set dictionary that shared among optimization passes."""
 
 
-This is sort of shared memory space among passes.
-"""
+from dataclasses import dataclass, field
+from enum import IntEnum
+
+from .exceptions import PassManagerError
 
 
 class PropertySet(dict):
-    """A default dictionary-like object"""
+    """A default dictionary-like object."""
 
     def __missing__(self, key):
         return None
+
+
+class FencedPropertySet(PropertySet):
+    """A readonly property set that cannot be written via __setitem__."""
+
+    def __setitem__(self, key, value):
+        raise PassManagerError("The fenced PropertySet has the property __setitem__ protected.")
+
+
+class RunState(IntEnum):
+    """Allowed values for the result of a pass execution."""
+
+    FAIL = 0
+    SUCCESS = 1
+    SKIP = 2
+
+
+@dataclass
+class PassState:
+    """Collection of compilation status.
+
+    This information is initialized in the pass manager instance at construction time,
+    and recursively passed to flow controllers at running time.
+    Passes can mutate the information in this data set.
+    """
+
+    property_set: PropertySet = field(default_factory=PropertySet)
+    """Property set dictionary that shares intermediate information across passes."""
+
+    count: int = 0
+    """Current number of pass execution."""
+
+    completed_passes: set = field(default_factory=set)
+    """Passes already run that have not been invalidated."""
+
+    previous_run: RunState = 0
+    """Status of the latest pass run."""
```

Comment:
This is one of my concerns about the `enum` I wrote above: if we're going to have an `enum`, it's not clear to me that `0` should be an allowed value.  If anything, I'd actually expect that zero should be the _success_ case (cf Unix-style return values from programs), but if we enforced that this needs to be spelt `previous_run = RunState.FAIL` then there'd be no chance of confusion.

Reply:
Done in [5034cc4](https://github.com/Qiskit/qiskit/pull/10127/commits/5034cc49d2767ec1ef13ac526609af38d15119bb)